ast-nodes/statement.dart

```diff
@@ -1,9 +1,36 @@
+import '../syntax-error.dart';
 import 'node.dart';
 import '../lexer.dart';
+import 'while-loop.dart';
+import 'for-loop.dart';
+import 'if-statement.dart';
+import 'routine-call.dart';
+import 'assignment.dart';
 
 /// An abstract statement.
 abstract class Statement implements Node {
   factory Statement.parse(Iterable<Token> tokens) {
+    if (tokens.isEmpty) {
+      throw SyntaxError(tokens?.first, 'Expected a statement');
+    }
 
+    if (tokens.first.value == 'while')
+      return WhileLoop.parse(tokens);
+    else if (tokens.first.value == 'for')
+      return ForLoop.parse(tokens);
+    else if (tokens.first.value == 'if') return IfStatement.parse(tokens);
```

Comment:
Please, try to keep all your branches indented identically. The code like:
```python
if stuff:
  indented_branch()
else: unindented_branch()
```
is significantly harder to read (that applies not only to this file, but to several files in this PR, go through them all). Also I'd really appreciate if you used braces (`{}`) everywhere, even if you don't need them. Not only does this promote unity in code and allow you to add more statements to the `if` easily, but it also helps to avoid various pitfalls like the hanging `else`. Braces = good.